patches/server/0739-Add-CompostItemEvent-and-EntityCompostItemEvent.patch

```diff
@@ -0,0 +1,101 @@
+         if (i < 7 && ComposterBlock.COMPOSTABLES.containsKey(itemstack.getItem())) {
+             // CraftBukkit start
+             double rand = worldserver.getRandom().nextDouble();
+-            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand);
++            BlockState iblockdata1 = ComposterBlock.addItem(iblockdata, DummyGeneratorAccess.INSTANCE, blockposition, itemstack, rand, entity); // Paper
+             if (iblockdata == iblockdata1 || org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, blockposition, iblockdata1).isCancelled()) {
+                 return iblockdata;
+             }
+-            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand);
++            // Paper start
++            iblockdata1 = ComposterBlock.addItem(iblockdata, (LevelAccessor) worldserver, blockposition, itemstack, rand, entity);
++            if (iblockdata1 == null) {
++                return iblockdata;
++            }
++            // Paper end
+             // CraftBukkit end
+ 
+             itemstack.shrink(1);
+@@ -291,16 +301,42 @@ public class ComposterBlock extends Block implements WorldlyContainerHolder {
+     }
+ 
+     static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item) {
+-        // CraftBukkit start
+-        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble());
++        return ComposterBlock.addItem(state, world, pos, item, null);
++    }
++    // Paper start
++    static BlockState addItem(BlockState state, LevelAccessor world, BlockPos pos, ItemStack item, @Nullable Entity entity) {
++        return ComposterBlock.addItem(state, world, pos, item, world.getRandom().nextDouble(), entity);
+     }
++    // Paper end
+ 
+     static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand) {
++        // Paper start
++        return ComposterBlock.addItem(iblockdata, generatoraccess, blockposition, itemstack, rand, null);
++    }
++
++    static BlockState addItem(BlockState iblockdata, LevelAccessor generatoraccess, BlockPos blockposition, ItemStack itemstack, double rand, @Nullable Entity entity) {
++        // Paper end
+         // CraftBukkit end
+         int i = (Integer) iblockdata.getValue(ComposterBlock.LEVEL);
+         float f = ComposterBlock.COMPOSTABLES.getFloat(itemstack.getItem());
+ 
+-        if ((i != 0 || f <= 0.0F) && rand >= (double) f) {
++        // Paper start
++        com.destroystokyo.paper.event.block.CompostItemEvent event = null;
++        if (generatoraccess != org.bukkit.craftbukkit.util.DummyGeneratorAccess.INSTANCE) {
++            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f); // taken from if conditional below
```

Comment:
This comment is wrong now since its not really "below" anymore.

```suggestion
+            boolean willRaiseLevel = !((i != 0 || f <= 0.0F) && rand >= (double) f);
```